### PR TITLE
Add new functions/methods to featurize module

### LINKED ISSUE #329: Update featurizer module with more options
Add new features like icohps weighted by reduced masses, bond specific dataframe, possibility to work with kwargs

### PR DESCRIPTION
**Pull request by naik-aakash**
Closes #329 

# Changes
- [x] Add ultiliy functions to compute reduced masses, electronegativity and sort dict by values
- [x] Add the possibility to modify default lobsterpy kwargs in featurizers
- [x] Add ICOXX featurizer with BWDF functionality
- [x] Get BWDF statistical features
- [x] Investigate incosistencies in translation vectors wrt bond distances in ICOXXLIST.lobster (result :-> only inconsistent if POSCAR is read, it is consistent when CONTCAR/POSCAR.lobster/POSCAR.lobster.vasp is read)
- [x] Make CONTCAR as default file to read in automatic analysis and featurizers (translations are then consistent with lobster files)
- [x]  Add tests

### CODE REVIEW

**Comment by naik-aakash** [comment]
Hi @JaGeo, there seems to be some issue with coverage artificats not being generated and uploaded in CI, I have not been able to pinpoint the cause yet.  Will check again in detail tomorrow. Leaving this as a comment only for my reference here.

**Comment by naik-aakash** [comment]
Could be simply bug associated with a new release of pytest today. Will have to check downgrading it

**Comment by JaGeo** [comment]
@naik-aakash sure. Don't worry!

**Comment by naik-aakash** [comment]
> @naik-aakash sure. Don't worry!

issue was with upload artificats update ignoring hidden files from being uploaded. It should be fixed now in PR #331

File: src/lobsterpy/featurize/utils.py [hunk 1]
```diff
@@ -91,3 +92,40 @@ def get_structure_path(lobster_path: Path) -> Path:
...
+    """
+    Compute reduced mass between a pair of atoms.
+
+    :param atom_pair: list of atomic species symbols in string
+
+    :return: reduced mass
+    """
+    atom1 = element(atom_pair[0])
+    atom2 = element(atom_pair[1])
+    return (atom1.atomic_weight * atom2.atomic_weight) / (atom1.atomic_weight + atom2.atomic_weight)
```
[the remark this answers is not in the document]
Comment (naik-aakash) [reply src/lobsterpy/featurize/utils.py:107]: I now inlcude the mean, mean, max stats per bond type as per the suggestion

**Comment by naik-aakash** [comment]
Hi @JaGeo , posting here results of some tests I did to verify the implementation for BWDF

*Note*: In all the tests, a supercell of 2x2x2 is constructed 

## ZnSb Structure : [mp-753](https://next-gen.materialsproject.org/materials/mp-753)
bwdf_Sb-Sb_ZnSb
bwdf_Sb-Zn_ZnSb
bwdf_summed_ZnSb
bwdf_Zn-Zn_ZnSb

## NaCl 

bwdf_Cl-Cl_NaCl
bwdf_Cl-Na_NaCl
bwdf_Na-Na_NaCl
bwdf_summed_NaCl

## Si3N4 : [mp-11607](https://next-gen.materialsproject.org/materials/mp-11607)
bwdf_N-N_Si3N4
bwdf_N-Si_Si3N4
bwdf_Si-Si_Si3N4
bwdf_summed_Si3N4

## AlN : [mp-661](https://next-gen.materialsproject.org/materials/mp-661)

bwdf_Al-Al_AlN
bwdf_Al-N_AlN
bwdf_N-N_AlN
bwdf_summed_AlN

**Comment by naik-aakash** [comment]
Just for NaCl and AlN, I do not have exactly overlapping peaks. But seems to work in other cases. Can't seem to figure out what could be the reason.

**Comment by JaGeo** [comment]
@naik-aakash can you provide the structures for both AlN cases and both NaCl cases here?

**Comment by naik-aakash** [comment]
> @naik-aakash can you provide the structures for both AlN cases and both NaCl cases here? And icohps for all Al-Al and cl-cl.interactions?

cannot upload compressed files here, will share the calc files on email.

**Comment by naik-aakash** [comment]
Hi @JaGeo , Below you can find the outputs for the same structures with updated implementation, which now works fine for primitive and supercells

### NaCl
bwdf_Cl-Cl_NaCl_new
bwdf_Cl-Na_NaCl_new
bwdf_Na-Na_NaCl_new
bwdf_summed_NaCl_new

### AlN
bwdf_Al-Al_AlN_new
bwdf_Al-N_AlN_new
bwdf_N-N_AlN_new
bwdf_summed_AlN_new

### Si3N4
bwdf_N-N_Si3N4_new
bwdf_N-Si_Si3N4_new
bwdf_Si-Si_Si3N4_new
bwdf_summed_Si3N4_new

### ZnSb
bwdf_Sb-Sb_ZnSb_new
bwdf_Sb-Zn_ZnSb_new
bwdf_summed_ZnSb_new
bwdf_Zn-Zn_ZnSb_new

**Comment by JaGeo** [comment]
Very good. I think you can test some more cases and see if your implementation can cover all of those

**Comment by QuantumChemist** [comment]
Oh, I replied to the wrong comment

**Comment by naik-aakash** [comment]
Hi @JaGeo  and @QuantumChemist , I did some more tests on different kinds of structures for supercell and primitive comparisons. Just posting here some results of summed BWDF I got from these comparisons for this structures. It seems to work in these cases as well.

## orthorhombic : [mp-10381](https://next-gen.materialsproject.org/materials/mp-10381/)
ortho_bwdf_summed_K2CuSb_new

## monoclinic : [mp-10408](https://next-gen.materialsproject.org/materials/mp-10408/)
mono_bwdf_summed_K2CN2_new

## hexagonal : [mp-10614](https://next-gen.materialsproject.org/materials/mp-10614/)
hexa_bwdf_summed_SrLiP_new

## triclinic : [mp-14983](https://next-gen.materialsproject.org/materials/mp-14983/)
tri_bwdf_summed_Si4P4Ru_new

**Comment by naik-aakash** [comment]
> Good, I would recommend testing your implementation on your complete dataset and check if there are any failures (does the algorithm find all ICOHPs in the ICOHP list?)

Will check this soon 😄

**Comment by QuantumChemist** [comment]
> Hi @JaGeo and @QuantumChemist , I did some more tests on different kinds of structures for supercell and primitive comparisons. Just posting here some results of summed BWDF I got from these comparisons for this structures. It seems to work in these cases as well.
> 
> ## orthorhombic : [mp-10381](https://next-gen.materialsproject.org/materials/mp-10381/)
> ortho_bwdf_summed_K2CuSb_new
> 
> ## monoclinic : [mp-10408](https://next-gen.materialsproject.org/materials/mp-10408/)
> mono_bwdf_summed_K2CN2_new
> 
> ## hexagonal : [mp-10614](https://next-gen.materialsproject.org/materials/mp-10614/)
> hexa_bwdf_summed_SrLiP_new
> 
> ## triclinic : [mp-14983](https://next-gen.materialsproject.org/materials/mp-14983/)
> tri_bwdf_summed_Si4P4Ru_new

Fantastic!

File: src/lobsterpy/featurize/core.py [hunk 2]
```diff
@@ -1192,7 +1192,7 @@ def __init__(
         :param path_to_icoxxlist: path to ICOXXLIST.lobster
         :param path_to_structure: path to POSCAR
         :param bin_width: bin width for the BWDF
-        :param interactions_tol: tolerance for interactions
+        :param interactions_tol: numerical tolerance considered for interactions to be zero
```
Comment (QuantumChemist) [inline src/lobsterpy/featurize/core.py:1195]:
```suggestion
        :param interactions_tol: numerical tolerance considered for interactions to be insignificant
```

File: src/lobsterpy/featurize/core.py [hunk 3]
```diff
@@ -1146,3 +1132,462 @@ def get_fingerprint_df(
...
+        else:
+            ids = Path(self.path_to_icoxxlist).parent.name
+            df = pd.DataFrame(index=[ids], columns=column_names)
+
+        for icoxx_weight, col in zip(site_bwdf[f"{site_index}"]["icoxx_binned"], column_names):
+            df.loc[ids, col] = icoxx_weight
+
+        return df
+
+    def get_stats_df(self, ids: str | None = None) -> pd.DataFrame:
```
Comment (kaueltzen) [inline src/lobsterpy/featurize/core.py:1552]: Hi @naik-aakash another BWDF featurization that avoids many zero entries could be (a) a list of distances ordered by BWDF value, descending or (b) a list of BWDF values ordered by distances, ascending. If you are interested, i'm happy to add that.

Comment (JaGeo) [reply src/lobsterpy/featurize/core.py:1552]: I think this sounds like a very good idea.

Comment (naik-aakash) [reply src/lobsterpy/featurize/core.py:1574]: Thanks @kaueltzen for this suggestion. Could be nice to also test this out. I have now pushed the changes to repo, you can add it to this branch

**Review by QuantumChemist — COMMENTED** [review]
Allen is a name and therefore written with capital A 😃

**Comment by naik-aakash** [comment]
Hi @JaGeo , I think this PR could be merged now

File: src/lobsterpy/featurize/core.py [hunk 4]
```diff
@@ -1146,3 +1132,530 @@ def get_fingerprint_df(
...
+            bins=np.arange(0, self.max_length + self.bin_width, self.bin_width),
+            density=False,
+        )
+
+        return {
+            "ref_rdf_data": ref_rdf_data,
+            "input_icoxx_list": input_icoxx_list,
+            "mapped_icoxx_data": mapped_icoxx_data,
+            "missing_interactions": missing_interactions,
+            "wasserstein_dist_to_rdf": wasserstein_distance(rdf_hist, icoxx_hist),
```
Comment (JaGeo) [inline src/lobsterpy/featurize/core.py:1325]: Is this documented somewhere? I don't think I can see a description .

File: src/lobsterpy/featurize/core.py [hunk 5]
```diff
@@ -1146,3 +1132,530 @@ def get_fingerprint_df(
...
+        return bwdf
+
+    def get_binned_bwdf_df(self, ids: str | None = None) -> pd.DataFrame:
+        """Return a pandas dataframe with computed BWDF features as columns.
+
+        Args:
+            ids: set index name in the pandas dataframe. Default is None.
+
+        Returns:
+            A pandas dataframe object with BWDF features
```
Comment (JaGeo) [inline src/lobsterpy/featurize/core.py:1532]: Again, we need more documentation here

File: src/lobsterpy/featurize/core.py (hunk 4 above)
Comment (naik-aakash) [reply src/lobsterpy/featurize/core.py:1330]: Added detailed descriptions now in doc-strings. Thanks for catching it

File: src/lobsterpy/featurize/core.py (hunk 5 above)
Comment (naik-aakash) [reply src/lobsterpy/featurize/core.py]: Added detailed descriptions now in doc-strings. Thanks for catching it

File: README.md [hunk 6]
```diff
@@ -87,9 +87,13 @@ Please cite our papers:
 * A. A. Naik, K. Ueltzen, C. Ertural, A. J. Jackson, J. George, *Journal of Open Source Software* **2024**, *9*, 6286. [https://joss.theoj.org/papers/10.21105/joss.06286](https://joss.theoj.org/papers/10.21105/joss.06286).
 * J. George, G. Petretto, A. Naik, M. Esters, A. J. Jackson, R. Nelson, R. Dronskowski, G.-M. Rignanese, G. Hautier, *ChemPlusChem* **2022**, *87*, e202200123. [https://doi.org/10.1002/cplu.202200123](https://doi.org/10.1002/cplu.202200123) (Information on the methodology of the automatic analysis)
 
+If you use any of the following Featurizers, also cite the respective papers:
+
+* `FeaturizeCharges`: R. Nelson, C. Ertural, P. C. Müller, R. Dronskowski, in Comprehensive Inorganic Chemistry III, *Elsevier*, **2023**, pp. 141–201. [https://doi.org/10.1016/B978-0-12-823144-9.00120-5](https://doi.org/10.1016/B978-0-12-823144-9.00120-5)
+* `FeaturizeIcoxxlist`: V. L. Deringer, W. Zhang, M. Lumeij, S. Maintz, M. Wuttig, R. Mazzarello, R. Dronskowski, *Angewandte Chemie International Edition 2014*, *53*, 10817–10820. [https://doi.org/10.1002/anie.201404223](https://doi.org/10.1002/anie.201404223)
```
Comment (JaGeo) [inline README.md:93]: It would be nice to add all formulas that have been implemented.

Comment (naik-aakash) [reply README.md:93]: Only BWDF is missing, Will add it in here https://jageo.github.io/LobsterPy/fundamentals/index.html#features-curated-for-ml-via-featurizer, to keep it consistent

Comment (JaGeo) [reply README.md:93]: Yes please.

Comment (naik-aakash) [reply README.md:93]: Done

File: src/lobsterpy/cli.py [hunk 7]
```diff
@@ -561,7 +561,7 @@ def get_parser() -> argparse.ArgumentParser:
         ],
         help=(
             "Deliver a text description of the LOBSTER calc quality analysis. "
-            "Mandatory required files: POSCAR, POTCAR, lobsterout, lobsterin. "
+            "Mandatory required files: CONTCAR, POTCAR, lobsterout, lobsterin. "
```
Comment (kaueltzen) [inline src/lobsterpy/cli.py:564]: Would change it to "CONTCAR or other structure file compatible with --file-structure"

File: src/lobsterpy/featurize/utils.py [hunk 8]
```diff
@@ -76,13 +102,13 @@ def get_structure_path(lobster_path: Path) -> Path:
...
 
-    POSCAR is prioritized over POSCAR.lobster.
+    CONTCAR is prioritized over POSCAR.lobster.
 
     :param lobster_path: path to root LOBSTER calc directory
 
     :return: path to structure file
     """
-    for filename in ["POSCAR", "POSCAR.lobster", "POSCAR.lobster.vasp"]:
+    for filename in ["CONTCAR", "POSCAR.lobster", "POSCAR.lobster.vasp"]:
```
Comment (kaueltzen) [inline src/lobsterpy/featurize/utils.py:111]: add "POSCAR" as last option?

File: docs/tutorial/commandlineinterface.rst [hunk 9]
```diff
@@ -15,7 +15,7 @@ Creating input files
    
 
 With LobsterPy, these intricate details are handled with a single command. We need the standard VASP input files, i.e. 
-``INCAR, KPOINTS, POTCAR and POSCAR`` in the calculation directory. Once you have these files, one needs to run the following command:
+``INCAR, KPOINTS, POTCAR and CONTCAR`` in the calculation directory. Once you have these files, one needs to run the following command:
```
Comment (kaueltzen) [inline docs/tutorial/commandlineinterface.rst:18]: This needs to be adapted in the other .rst files as well, e.g. descriptionquality.rst

**Review by kaueltzen — COMMENTED** [review]
Maybe a warning if POSCAR is parsed anywhere would be the most elegant way?

File: src/lobsterpy/cli.py [hunk 10]
```diff
@@ -63,10 +63,10 @@ def get_parser() -> argparse.ArgumentParser:
...
         "-fstruct",
         "--file-structure",
-        default="POSCAR",
+        default="CONTCAR",
         dest="structure",
         type=Path,
-        help='path to structure file. Default is "POSCAR". '
+        help='path to structure file. Default is "CONTCAR". '
         'Can also read "POSCAR.lobster" file or any '
         'suitable file format supported by pymatgen "Structure.from_file" method.',
```
[the remark this answers is not in the document]
Comment (naik-aakash) [reply src/lobsterpy/cli.py:71]: Sounds good

File: src/lobsterpy/featurize/utils.py (hunk 8 above)
Comment (naik-aakash) [reply src/lobsterpy/featurize/utils.py]: Makes sense, yes

**Review by kaueltzen — APPROVED** [review]

**Review by JaGeo — APPROVED** [review]

**Pull request merged**